pilot/pkg/xds/ads.go

```diff
@@ -386,25 +386,19 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.Disc
 
 	if shouldUnsubscribe(request) {
 		log.Debugf("ADS:%s: UNSUBSCRIBE %s %s %s", stype, con.conID, request.VersionInfo, request.ResponseNonce)
-		con.proxy.Lock()
-		delete(con.proxy.WatchedResources, request.TypeUrl)
-		con.proxy.Unlock()
+		con.proxy.Unsubscribe(request.TypeUrl)
 		return false, emptyResourceDelta
 	}
 
-	con.proxy.RLock()
-	previousInfo := con.proxy.WatchedResources[request.TypeUrl]
-	con.proxy.RUnlock()
-
+	previousInfo := con.proxy.GetWatchedResource(request.TypeUrl)
 	// This can happen in two cases:
 	// 1. When Envoy starts for the first time, it sends an initial Discovery request to Istiod.
 	// 2. When Envoy reconnects to a new Istiod that does not have information about this typeUrl
 	// i.e. non empty response nonce.
 	// We should always respond with the current resource names.
 	if request.ResponseNonce == "" || previousInfo == nil {
 		log.Debugf("ADS:%s: INIT/RECONNECT %s %s %s", stype, con.conID, request.VersionInfo, request.ResponseNonce)
-		con.proxy.Lock()
-		con.proxy.WatchedResources[request.TypeUrl] = &model.WatchedResource{TypeUrl: request.TypeUrl, ResourceNames: request.ResourceNames}
+		con.proxy.AddOrUpdateWatchedResource(&model.WatchedResource{TypeUrl: request.TypeUrl, ResourceNames: request.ResourceNames})
```

Comment:
> Not sure it matters too much, but this is technically worse than before since we do 2 locks instead of 1?

Are n't locks same?

> I wonder if we should just take one big lock out on basically this whole function (starting on L393)

I think that seems better overall instead of two locks

Reply:
I mean its one lock but we do many lock+unlock calls instead of just one. This could actually in theory lead to inconsistent state

Reply:
yes. better to have a single lock here for sure

Reply:
actually, we donot need to.  the access to watchedResources is guaranteed to be in serial. The reason why we need lock is the debug interface allows read about it too.

Reply:
If we use one lock to protect the whole function, then we cannot make all the proxy's method concurrent safe, then we would go back to the previous statues, lock/unlock everywhere across ads handling.

Reply:
We can still have those methods in Proxy that abstracts lock and used every where else but in this specific function we can take a big lock and call update/read without locks. I am fine either way because this change is inline with what we have already.
@howardjohn  WDYT? If this looks good, please approve

Reply:
I think this introduces a regression. We should not make our locks so fine grained.

```go
for .. {
  lock()
  ...
  unlock()
}
```

Is not a desired pattern, but it is what is introduced here. They helpers in general are nice but we should make sure they are not making it easy to introduce undesired behaviors.

Reply:
can you elaborate the regression a little bit, so i can update

Reply:
Before
```go
lock()
for .. {
  ...
}
unlock()
```

After
```go
for .. {
  lock()
  ...
  unlock()
}
```

I think the before is preferred. Tbh everything probably ends up being correct enough (especially if you say the only concurrent access is debug interface) so I won't block this if it's not feasible to change, just seems a bit better

Reply:
OK, let me think about it. Will feed back later

Reply:
Looks at this commit  https://github.com/istio/istio/pull/48846/commits/016ffa7e5b7632130dafcce16ccfde5a4ca53d94 it is not good. We still donot make all the WatrchesResources ops atomic.

Locking too much critical section is not a good practice. FYI  https://en.wikipedia.org/wiki/Critical_section

Reply:
addressed, protect this snippet with one lock